File: backend/src/main/kotlin/com/casadosreclamos/dto/RequestTypeDto.kt
package com.casadosreclamos.dto

import com.casadosreclamos.exception.InvalidRequestTypeException
import com.casadosreclamos.model.request.RequestType
import com.casadosreclamos.model.request.OneFace as OneFaceRequest
import com.casadosreclamos.model.request.TwoFaces as TwoFacesRequest
import com.casadosreclamos.model.request.LeftShowcase as LeftRequest
import com.casadosreclamos.model.request.RightShowcase as RightRequest
import com.casadosreclamos.model.request.SimpleShowcase as SimpleRequest
import com.casadosreclamos.model.request.Showcase
import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo
import org.eclipse.microprofile.openapi.annotations.enums.SchemaType
import org.eclipse.microprofile.openapi.annotations.media.Schema

private const val PROPERTY = "type"

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = PROPERTY)
@JsonSubTypes(
    JsonSubTypes.Type(OneFace::class, name = "OneFace"),
    JsonSubTypes.Type(TwoFaces::class, name = "TwoFaces"),
    JsonSubTypes.Type(SimpleShowcase::class, name = "SimpleShowcase"),
    JsonSubTypes.Type(LeftShowcase::class, name = "LeftShowcase"),
    JsonSubTypes.Type(RightShowcase::class, name = "RightShowcase")
)
@Schema(
    type = SchemaType.OBJECT,
    discriminatorProperty = PROPERTY,
    oneOf = [OneFace::class, TwoFaces::class, SimpleShowcase::class, LeftShowcase::class, RightShowcase::class]
)
abstract class RequestTypeDto {
    companion object {
        fun from(request: RequestType): RequestTypeDto {
            return when (request) {
                is OneFaceRequest -> fromOneFace(request)
                is TwoFacesRequest -> fromTwoFaces(request)
                is SimpleRequest -> fromSimpleShowcase(request)
                is LeftRequest -> fromLeftShowcase(request)
                is RightRequest -> fromRightShowcase(request)
                else -> throw InvalidRequestTypeException()
            }
        }
    }
}

abstract class ShowcaseDto: RequestTypeDto() {
    lateinit var top: RequestSlotDto
    lateinit var bottom: RequestSlotDto
    lateinit var left: RequestSlotDto
    lateinit var right: RequestSlotDto
    lateinit var side: RequestSlotDto

    fun fromRequest(request: Showcase) {
        this.top = RequestSlotDto(request.top)
        this.bottom = RequestSlotDto(request.bottom)
        this.left = RequestSlotDto(request.left)
        this.right = RequestSlotDto(request.right)
        this.side = RequestSlotDto(request.side)
    }

    override fun toString(): String {
        return "TOP: ($top), BOTTOM: ($bottom), LEFT: ($left), RIGHT: ($right), SIDE: ($side)"
    }
}

class OneFace() : RequestTypeDto() {
    lateinit var cover: RequestSlotDto

    constructor(request: OneFaceRequest) : this() {
        this.cover = RequestSlotDto(request.cover)
    }

    override fun toString(): String {
        return "FACE: $cover"
    }
}

class TwoFaces() : RequestTypeDto() {
    lateinit var cover: RequestSlotDto

    lateinit var back: RequestSlotDto


    constructor(request: TwoFacesRequest) : this() {
        this.cover = RequestSlotDto(request.cover)
        this.back = RequestSlotDto(request.back)
    }

    override fun toString(): String {
        return "FRONT: $cover, BACK: $back"
    }
}

class SimpleShowcase(): RequestTypeDto() {
    lateinit var top: RequestSlotDto
    lateinit var bottom: RequestSlotDto
    lateinit var left: RequestSlotDto
    lateinit var right: RequestSlotDto

    constructor(request: SimpleRequest): this() {
        this.top = RequestSlotDto(request.top)
        this.bottom = RequestSlotDto(request.bottom)
        this.left = RequestSlotDto(request.left)
        this.right = RequestSlotDto(request.right)
    }
}

class LeftShowcase() : ShowcaseDto() {
    constructor(request: LeftRequest) : this() {
        this.fromRequest(request)
    }
}
class RightShowcase() : ShowcaseDto() {
    constructor(request: RightRequest) : this() {
        this.fromRequest(request)
    }
}

private fun fromOneFace(request: OneFaceRequest): OneFace {
    return OneFace(request)
}

private fun fromTwoFaces(request: TwoFacesRequest): TwoFaces {
    return TwoFaces(request)
}

private fun fromSimpleShowcase(request: SimpleRequest): SimpleShowcase {
    return SimpleShowcase(request)
}

private fun fromLeftShowcase(request: LeftRequest): LeftShowcase {
    return LeftShowcase(request)
}

private fun fromRightShowcase(request: RightRequest): RightShowcase {
    return RightShowcase(request)
}

File: frontend/src/models/dto/RequestTypeDto.ts
import RequestSlotDto from "@models/dto/RequestSlotDto";
import { RequestType } from "@models/dto/RequestType";

export class RequestTypeDto {
  type: RequestType | null = null;

  static from(obj: RequestTypeDto): RequestTypeDto {
    switch (obj.type as string) {
      case "OneFace":
        return new OneFace(obj as OneFace);
      case "TwoFaces":
        return new TwoFaces(obj as TwoFaces);
      case "LeftShowcase":
        return new LeftShowcase(obj as LeftShowcase);
      case "RightShowcase":
        return new RightShowcase(obj as RightShowcase);
      case "SimpleShowcase":
        return new SimpleShowcase(obj as SimpleShowcase);
      default:
        throw new Error(`Unknown request type: ${obj.type}`);
    }
  }

  getFinishings?(): string[][];
  getMaterials?(): string[];
  getMeasurements?(): number[][];
}

export class OneFace extends RequestTypeDto {
  cover: RequestSlotDto;

  constructor(obj: OneFace) {
    super();

    this.cover = new RequestSlotDto(obj.cover);
    this.type = RequestType.OneFace;
  }

  getMaterials(): string[] {
    return [this.cover.getMaterial()];
  }

  getFinishings(): string[][] {
    return [this.cover.getFinishings()];
  }

  getMeasurements(): number[][] {
    return [this.cover.getMeasurements()];
  }
}

export class TwoFaces extends RequestTypeDto {
  cover: RequestSlotDto;
  back: RequestSlotDto;

  constructor(obj: TwoFaces) {
    super();

    this.cover = new RequestSlotDto(obj.cover);
    this.back = new RequestSlotDto(obj.back);
    this.type = RequestType.TwoFaces;
  }

  getMaterials(): string[] {
    return [this.cover.getMaterial(), this.back.getMaterial()];
  }

  getFinishings(): string[][] {
    return [this.cover.getFinishings(), this.back.getFinishings()];
  }

  getMeasurements(): number[][] {
    return [this.cover.getMeasurements(), this.back.getMeasurements()];
  }
}

export class ShowcaseDto extends RequestTypeDto {
  top: RequestSlotDto;
  bottom: RequestSlotDto;
  left: RequestSlotDto;
  right: RequestSlotDto;
  side: RequestSlotDto | null = null;

  constructor(obj: ShowcaseDto) {
    super();

    this.top = new RequestSlotDto(obj.top);
    this.bottom = new RequestSlotDto(obj.bottom);
    this.left = new RequestSlotDto(obj.left);
    this.right = new RequestSlotDto(obj.right);

    if (obj.side) {
      this.side = new RequestSlotDto(obj.side);
    }
  }
}

export class LeftShowcase extends ShowcaseDto {
  type = RequestType.LeftShowcase;

  constructor(obj: LeftShowcase) {
    super(obj);
  }

  getMaterials(): string[] {
    return [
      this.top.getMaterial(),
      this.bottom.getMaterial(),
      this.left.getMaterial(),
      this.right.getMaterial(),
      this.side!!.getMaterial(),
    ];
  }

  getFinishings(): string[][] {
    return [
      this.top.getFinishings(),
      this.bottom.getFinishings(),
      this.left.getFinishings(),
      this.right.getFinishings(),
      this.side!!.getFinishings(),
    ];
  }

  getMeasurements(): number[][] {
    return [
      this.top.getMeasurements(),
      this.bottom.getMeasurements(),
      this.left.getMeasurements(),
      this.right.getMeasurements(),
      this.side!!.getMeasurements(),
    ];
  }
}

export class RightShowcase extends ShowcaseDto {
  type = RequestType.RightShowcase;

  constructor(obj: RightShowcase) {
    super(obj);
  }

  getMaterials(): string[] {
    return [
      this.top.getMaterial(),
      this.bottom.getMaterial(),
      this.left.getMaterial(),
      this.right.getMaterial(),
      this.side!!.getMaterial(),
    ];
  }

  getFinishings(): string[][] {
    return [
      this.top.getFinishings(),
      this.bottom.getFinishings(),
      this.left.getFinishings(),
      this.right.getFinishings(),
      this.side!!.getFinishings(),
    ];
  }

  getMeasurements(): number[][] {
    return [
      this.top.getMeasurements(),
      this.bottom.getMeasurements(),
      this.left.getMeasurements(),
      this.right.getMeasurements(),
      this.side!!.getMeasurements(),
    ];
  }
}

export class SimpleShowcase extends ShowcaseDto {
  type = RequestType.SimpleShowcase;

  constructor(obj: SimpleShowcase) {
    super(obj);
  }

  getMaterials(): string[] {
    return [
      this.top.getMaterial(),
      this.bottom.getMaterial(),
      this.left.getMaterial(),
      this.right.getMaterial(),
    ];
  }

  getFinishings(): string[][] {
    return [
      this.top.getFinishings(),
      this.bottom.getFinishings(),
      this.left.getFinishings(),
      this.right.getFinishings(),
    ];
  }

  getMeasurements(): number[][] {
    return [
      this.top.getMeasurements(),
      this.bottom.getMeasurements(),
      this.left.getMeasurements(),
      this.right.getMeasurements(),
    ];
  }
}
